src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -644,6 +655,133 @@ private static unsafe void Ssse3Decode(ref byte* srcBytes, ref byte* destBytes,
+
+            // Some constants.
+            Vector128<byte> vzero = Vector128.Create((byte)0);
+            Vector128<byte> v255 = Vector128.Create((byte)255U);
+            Vector128<byte> v16 = Vector128.Create((byte)16U);
+
+            byte* src =  srcBytes;
+            byte* dest = destBytes;
+
+            do
+            {
+                // Load 64 bytes of data and deinterleave the result.
+                // This is equivalent to a NEON LD4 instruction.
+                Vector128<byte> str0 = Vector128.LoadUnsafe(ref *src);
+                Vector128<byte> str1 = Vector128.LoadUnsafe(ref *src, 16);
+                Vector128<byte> str2 = Vector128.LoadUnsafe(ref *src, 32);
+                Vector128<byte> str3 = Vector128.LoadUnsafe(ref *src, 48);
+                Vector128<short> tmp0 = AdvSimd.Arm64.UnzipEven(str0.AsInt16(), str1.AsInt16());
+                Vector128<short> tmp1 = AdvSimd.Arm64.UnzipOdd(str0.AsInt16(), str1.AsInt16());
+                Vector128<short> tmp2 = AdvSimd.Arm64.UnzipEven(str2.AsInt16(), str3.AsInt16());
+                Vector128<short> tmp3 = AdvSimd.Arm64.UnzipOdd(str2.AsInt16(), str3.AsInt16());
+                str0 = AdvSimd.Arm64.UnzipEven(tmp0.AsByte(), tmp2.AsByte());
+                str1 = AdvSimd.Arm64.UnzipOdd(tmp0.AsByte(), tmp2.AsByte());
+                str2 = AdvSimd.Arm64.UnzipEven(tmp1.AsByte(), tmp3.AsByte());
+                str3 = AdvSimd.Arm64.UnzipOdd(tmp1.AsByte(), tmp3.AsByte());
+
+                // Table lookup on each 16 bytes.
+                str0 = AdvSimdTbx8Byte(v255, dec_lut0, dec_lut1, dec_lut2, dec_lut3, dec_lut4, dec_lut5, dec_lut6, dec_lut7, str0, v16);
+                str1 = AdvSimdTbx8Byte(v255, dec_lut0, dec_lut1, dec_lut2, dec_lut3, dec_lut4, dec_lut5, dec_lut6, dec_lut7, str1, v16);
+                str2 = AdvSimdTbx8Byte(v255, dec_lut0, dec_lut1, dec_lut2, dec_lut3, dec_lut4, dec_lut5, dec_lut6, dec_lut7, str2, v16);
+                str3 = AdvSimdTbx8Byte(v255, dec_lut0, dec_lut1, dec_lut2, dec_lut3, dec_lut4, dec_lut5, dec_lut6, dec_lut7, str3, v16);
+
+                // Check for invalid input, any value larger than 63.
+                Vector128<byte> classified0 = AdvSimd.Arm64.MaxPairwise(str0, str1);
+                Vector128<byte> classified1 = AdvSimd.Arm64.MaxPairwise(str2, str3);
+                Vector128<byte> maxChars = AdvSimd.Arm64.MaxPairwise(classified0, classified1);
+                if ((maxChars.AsUInt64().ToScalar() & 0xc0c0c0c0c0c0c0c0) != 0)
+                    break;
+
+                // Compress each four bytes into three.
+                Vector128<byte> dec0 = Vector128.BitwiseOr(Vector128.ShiftLeft(str0, 2), Vector128.ShiftRightLogical(str1, 4));
+                Vector128<byte> dec1 = Vector128.BitwiseOr(Vector128.ShiftLeft(str1, 4), Vector128.ShiftRightLogical(str2, 2));
+                Vector128<byte> dec2 = Vector128.BitwiseOr(Vector128.ShiftLeft(str2, 6), str3);
```

Comment:
jit doesn't do instruction selection so if you want your shifts to be side-by-side for better pipelining you need to extract them to temp locals, e.g.:
```csharp
var sl0 = Vector128.ShiftLeft(str0, 2);
var sl1 = Vector128.ShiftLeft(str1, 4);
var sl2 = Vector128.ShiftLeft(str2, 6);
var sr1 = Vector128.ShiftRightLogical(str1, 4);
var sr2 = Vector128.ShiftRightLogical(str2, 2);

Vector128<byte> dec0 = Vector128.BitwiseOr(sl0, sr1);
Vector128<byte> dec1 = Vector128.BitwiseOr(sl1, sr2);
Vector128<byte> dec2 = Vector128.BitwiseOr(sl2, str3);
```

not sure it matters much in terms of perf
```

Reply:
Another way to do this might be to treat the vector register as a single value and do something fancy with a masks and a single value shift. Not quite sure how that would look. Might get a some performance, but it'd be messy.